File: .skills/openclaw-skills/skills/cp7553479/seedream-image-generation/scripts/seedream.js
const https = require('https');
const fs = require('fs');
const path = require('path');

/**
 * Converts a local image file to base64 data URI.
 * @param {string} imagePath - Path to the image file.
 * @returns {string} Base64 data URI (data:image/xxx;base64,...)
 */
function imageToBase64DataURI(imagePath) {
    const mimeTypes = {
        '.jpg': 'image/jpeg',
        '.jpeg': 'image/jpeg',
        '.png': 'image/png',
        '.gif': 'image/gif',
        '.webp': 'image/webp',
        '.bmp': 'image/bmp',
        '.tiff': 'image/tiff'
    };
    const ext = path.extname(imagePath).toLowerCase();
    const mime = mimeTypes[ext] || 'image/jpeg';
    const data = fs.readFileSync(imagePath);
    const base64 = data.toString('base64');
    return `data:${mime};base64,${base64}`;
}

/**
 * Generates an image using the Volcengine Seedream API.
 *
 * @param {Object} params - The inputs for the generation request.
 * @param {string} params.prompt - The text description of the image to generate. Required.
 * @param {string} [params.model] - The endpoint ID of the model. If missing, uses SEEDREAM_ENDPOINT_ID.
 * @param {string} [params.api_key] - Custom API key to use. Overrides environment variable.
 * @param {string} [params.size] - Target image size. Can be a resolution class (e.g., '2K', '3K') or explicit dimensions (e.g., '2048x2048').
 * @param {boolean} [params.watermark=false] - Whether to include an AI watermark.
 * @param {Object} [params.optimize_prompt_options] - Options to auto-optimize the prompt.
 * @param {Array} [params.tools] - A list of tools for the model to use (e.g., [{"type": "web_search"}]).
 * @param {string} [params.output_format] - Output image format (e.g., 'png', 'jpeg').
 * @param {string} [params.sequential_image_generation] - Sequential image generation strategy (e.g., 'auto').
 * @param {string} [params.download_dir] - Directory to save generated images.
 * @param {string|Array<string>} [params.image] - Local image path or array of local image paths to use as input (for I2I).
 *
 * @returns {Promise<Object>} A promise resolving to the API response object.
 *
 * Return value format (on success):
 * {
 *   "data": [
 *     {
 *       "url": "https://...",
 *       "size": "2048x2048"
 *     }
 *   ],
 *   "input_params": { ... },
 *   "usage": {
 *       "generated_images": 1,
 *       ...
 *   }
 * }
 *
 * Return value format (on failure):
 * {
 *   "error": "<Error Details or HTTP Code>",
 *   "message": "<Optional Server Message>"
 * }
 */
async function generateImage(params) {
    console.log('[seedream] generateImage input:', JSON.stringify({
        ...params,
        image: params.image !== undefined ? params.image : null
    }));
    const modelId = params.model || "doubao-seedream-5-0-260128";

    const payload = {
        model: modelId,
        prompt: params.prompt,
        watermark: params.watermark !== undefined && params.watermark !== null ? params.watermark : false
    };

    if (params.size !== undefined && params.size !== null) {
        payload.size = params.size;
    }
    if (params.optimize_prompt_options !== undefined && params.optimize_prompt_options !== null) {
        payload.optimize_prompt_options = params.optimize_prompt_options;
    }
    if (params.tools !== undefined && params.tools !== null) {
        payload.tools = params.tools;
    }
    if (params.output_format !== undefined && params.output_format !== null) {
        payload.output_format = params.output_format;
    }
    if (params.sequential_image_generation !== undefined && params.sequential_image_generation !== null) {
        payload.sequential_image_generation = params.sequential_image_generation;
    }
    if (params.response_format !== undefined && params.response_format !== null) {
        payload.response_format = params.response_format;
    }

    // Convert local image path(s) to base64 data URIs for I2I.
    if (params.image !== undefined && params.image !== null) {
        const inputImages = Array.isArray(params.image) ? params.image : [params.image];
        const imageDataUris = [];
        for (const imgPath of inputImages) {
            if (!fs.existsSync(imgPath)) {
                const result = { error: "Image not found", message: imgPath };
                console.error('[seedream] error:', JSON.stringify(result));
                return result;
            }
            try {
                const dataUri = imageToBase64DataURI(imgPath);
                imageDataUris.push(dataUri);
                console.log(`[seedream] loaded image: ${imgPath}`);
            } catch (e) {
                const result = { error: "Failed to load image", message: `${imgPath}: ${e.message}` };
                console.error('[seedream] error:', JSON.stringify(result));
                return result;
            }
        }

        if (imageDataUris.length === 1) {
            payload.image = imageDataUris[0];
        } else {
            payload.image = imageDataUris;
        }
        console.log(`[seedream] total images: ${imageDataUris.length}`);
    }

    const downloadImage = (url, dest) => {
        return new Promise((resolve, reject) => {
            const file = fs.createWriteStream(dest);
            https.get(url, (response) => {
                response.pipe(file);
                file.on('finish', () => {
                    file.close(resolve);
                });
            }).on('error', (err) => {
                fs.unlink(dest, () => { });
                reject(err.message);
            });
        });
    };

    const apiKey = process.env.SEEDREAM_API_KEY;
    if (!apiKey) {
        return { error: "Require SEEDREAM_API_KEY in environment" };
    }

    const baseUrl = process.env.SEEDREAM_BASE_URL || 'https://ark.cn-beijing.volces.com/api/v3/images/generations';
    const parsedUrl = new URL(baseUrl);

    console.log('[seedream] request:', JSON.stringify({
        url: baseUrl,
        payload: {
            ...payload,
            image: payload.image !== undefined ? '<base64_data_redacted>' : undefined
        }
    }));

    const dataString = JSON.stringify(payload);

    const options = {
        hostname: parsedUrl.hostname,
        port: parsedUrl.port || 443,
        path: parsedUrl.pathname + parsedUrl.search,
        method: 'POST',
        headers: {
            'Authorization': `Bearer ${apiKey}`,
            'Content-Type': 'application/json',
            'Content-Length': Buffer.byteLength(dataString)
        }
    };

    return new Promise((resolve) => {
        const req = https.request(options, (res) => {
            let resData = '';
            res.on('data', (chunk) => resData += chunk);
            res.on('end', async () => {
                try {
                    const resJson = JSON.parse(resData);
                    console.log('[seedream] response:', JSON.stringify(resJson));

                    if (resJson.data) {
                        resJson.input_params = payload;

                        if (params.download_dir) {
                            const saveDir = params.download_dir;
                            if (!fs.existsSync(saveDir)) {
                                fs.mkdirSync(saveDir, { recursive: true });
                            }

                            for (let i = 0; i < resJson.data.length; i++) {
                                const item = resJson.data[i];
                                if (item.url) {
                                    try {
                                        const ext = payload.output_format === 'png' ? '.png' : '.jpeg';
                                        const filename = `image_${resJson.created || Date.now()}_${i}${ext}`;
                                        const destPath = path.join(saveDir, filename);
                                        await downloadImage(item.url, destPath);
                                        item.local_path = destPath;
                                    } catch (err) {
                                        item.local_path_error = err;
                                    }
                                }
                            }
                        }
                    }
                    console.log('[seedream] output:', JSON.stringify(resJson));
                    resolve(resJson);
                } catch (e) {
                    const result = { error: "Parse error", message: resData };
                    console.error('[seedream] error:', JSON.stringify(result));
                    resolve(result);
                }
            });
        });

        req.on('error', (e) => {
            const result = { error: e.message };
            console.error('[seedream] error:', JSON.stringify(result));
            resolve(result);
        });
        req.write(dataString);
        req.end();
    });
}

if (require.main === module) {
    const args = process.argv.slice(2);
    const params = {};
    for (let i = 0; i < args.length; i++) {
        if (args[i].startsWith('--')) {
            const key = args[i].substring(2);
            params[key] = args[i + 1];
            i++;
        }
    }

    if (params.watermark === "true") params.watermark = true;
    if (params.watermark === "false") params.watermark = false;

    if (params.optimize_prompt_options) {
        try {
            params.optimize_prompt_options = JSON.parse(params.optimize_prompt_options);
        } catch (e) {
            console.error("Error: --optimize_prompt_options must be a valid JSON string");
            process.exit(1);
        }
    }

    if (params.tools) {
        try {
            params.tools = JSON.parse(params.tools);
        } catch (e) {
            console.error("Error: --tools must be a valid JSON string");
            process.exit(1);
        }
    }

    if (params.image) {
        if (params.image.trim().startsWith('[')) {
            try {
                params.image = JSON.parse(params.image);
                if (!Array.isArray(params.image)) {
                    console.error("Error: --image must be a local path or a JSON array of paths");
                    process.exit(1);
                }
            } catch (e) {
                console.error("Error: --image must be a local path or a valid JSON string array");
                process.exit(1);
            }
        }
    }

    if (!params.prompt) {
        console.error("Usage: node seedream.js --prompt <prompt> [--model <model>] [--size <size>] [--watermark <true/false>] [--output_format <format>] [--sequential_image_generation <mode>] [--tools <json>] [--optimize_prompt_options <json>] [--image <path|json_array>] [--download_dir <dir>]");
        process.exit(1);
    }

    generateImage(params).then(res => console.log(JSON.stringify(res, null, 2)));
}

module.exports = { generateImage };

File: .skills/openclaw-skills/skills/cp7553479/seedream-image-generation/scripts/seedream.py
import os
import sys
import json
import mimetypes
import urllib.request
import urllib.error
import argparse
import uuid
import base64

def generate_image(
    prompt,
    model=None,
    size=None,
    watermark=None,
    optimize_prompt_options=None,
    tools=None,
    output_format=None,
    sequential_image_generation=None,
    download_dir=None,
    image=None,
    response_format=None
):
    """
    Generates an image using the Volcengine Seedream API.
    
    Args:
        prompt (str): The text description of the image to generate. Required.
        model (str): The target model endpoint ID (e.g., ep-xxxx). Required.
        size (str, optional): Target image size. Can be a resolution class (e.g., '2K', '3K') 
                              or explicit dimensions (e.g., '2048x2048'). Defaults to None.
        watermark (bool, optional): Whether to add an AI watermark. Defaults to False.
        optimize_prompt_options (dict, optional): Options to optimize the prompt (e.g., {"mode": "standard"}). Defaults to None.
        tools (list, optional): A list of tools for the model to use (e.g., [{"type": "web_search"}]).
        output_format (str, optional): Output image format (e.g., 'png', 'jpeg', 'webp'). Defaults to None.
        sequential_image_generation (str, optional): Sequential image generation strategy (e.g., 'auto'). Defaults to None.
        image (str | list[str], optional): Local image path or image path list for I2I. Defaults to None.


        
    Returns:
        dict: The JSON response dictionary.
        
        Successful Return Value Format:
        {
          "data": [
            {
              "url": "https://...",
              "size": "2048x2048" # Included depending on the model capability
            }
          ],
          "usage": { 
              "generated_images": 1,
              ...
          }
        }
        
        Failure Return Value Format:
        {
          "error": "<Error Code or Details>",
          "message": "<Optional detailed message from server>"
        }
    """
    # Retrieve API key
    key = os.environ.get("SEEDREAM_API_KEY")
    if not key:
        err = {"error": "Require SEEDREAM_API_KEY in environment"}
        print("[seedream] error:", json.dumps(err, ensure_ascii=False))
        return err
    
    model_id = model or "doubao-seedream-5-0-260128"

    headers = {
        "Authorization": f"Bearer {key}",
    }

    image_input = [] if image is None else (image if isinstance(image, list) else [image])
    print("[seedream] generate_image input:", json.dumps({
        "prompt": prompt,
        "model": model,
        "size": size,
        "watermark": watermark,
        "optimize_prompt_options": optimize_prompt_options,
        "tools": tools,
        "output_format": output_format,
        "sequential_image_generation": sequential_image_generation,
        "download_dir": download_dir,
        "image": image,
        "response_format": response_format
    }, ensure_ascii=False))

    # Use SEEDREAM_BASE_URL if provided, else use default generations endpoint
    url = os.environ.get("SEEDREAM_BASE_URL", "https://ark.cn-beijing.volces.com/api/v3/images/generations")

    payload = {
        "model": model_id,
        "prompt": prompt,
        "watermark": watermark if watermark is not None else False
    }

    if size is not None:
        payload["size"] = size
    if optimize_prompt_options is not None:
        payload["optimize_prompt_options"] = optimize_prompt_options
    if tools is not None:
        payload["tools"] = tools
    if output_format is not None:
        payload["output_format"] = output_format
    if sequential_image_generation is not None:
        payload["sequential_image_generation"] = sequential_image_generation
    if response_format is not None:
        payload["response_format"] = response_format

    # Convert local image path(s) to base64 data URIs for I2I.
    if len(image_input) > 0:
        image_data_uris = []
        for img_path in image_input:
            if not os.path.exists(img_path):
                err = {"error": f"Image not found: {img_path}"}
                print("[seedream] error:", json.dumps(err, ensure_ascii=False))
                return err

            mime, _ = mimetypes.guess_type(img_path)
            if not mime:
                mime = "image/jpeg"

            with open(img_path, "rb") as f:
                img_data = f.read()

            b64_data = base64.b64encode(img_data).decode("utf-8")
            data_uri = f"data:{mime};base64,{b64_data}"
            image_data_uris.append(data_uri)
            print(f"[seedream] loaded image: {img_path} ({len(img_data)} bytes, mime: {mime})")

        # Support single image (string) or multiple images (array)
        if len(image_data_uris) == 1:
            payload["image"] = image_data_uris[0]
        else:
            payload["image"] = image_data_uris

        print(f"[seedream] total images: {len(image_data_uris)}")

    headers["Content-Type"] = "application/json"
    print("[seedream] request:", json.dumps({
        "url": url,
        "headers": {k: ("<redacted>" if k.lower() == "authorization" else v) for k, v in headers.items()},
        "payload": {**payload, "image": "<base64_data_redacted>" if "image" in payload else None}
    }, ensure_ascii=False))

    req = urllib.request.Request(url, data=json.dumps(payload).encode("utf-8"), headers=headers, method="POST")
    input_params_for_return = payload

    try:
        with urllib.request.urlopen(req) as response:
            res_json = json.loads(response.read().decode("utf-8"))
            print("[seedream] response:", json.dumps(res_json, ensure_ascii=False))

            # If successful, handle downloads if download_dir is provided
            if "data" in res_json:
                res_json["input_params"] = input_params_for_return

                if download_dir:
                    save_path = download_dir
                    if not os.path.exists(save_path):
                        os.makedirs(save_path)

                    output_format_used = input_params_for_return.get("output_format", "jpeg")

                    for item in res_json["data"]:
                        if "url" in item:
                            try:
                                file_ext = ".png" if output_format_used == "png" else ".jpeg"
                                filename = f"image_{int(res_json.get('created', 0))}_{res_json['data'].index(item)}{file_ext}"
                                local_file_path = os.path.join(save_path, filename)

                                urllib.request.urlretrieve(item["url"], local_file_path)
                                item["local_path"] = local_file_path
                            except Exception as download_err:
                                item["local_path_error"] = str(download_err)

            print("[seedream] output:", json.dumps(res_json, ensure_ascii=False))
            return res_json
    except urllib.error.HTTPError as e:
        err = {"error": e.code, "message": e.read().decode("utf-8")}
        print("[seedream] error:", json.dumps(err, ensure_ascii=False))
        return err
    except Exception as e:
        err = {"error": str(e)}
        print("[seedream] error:", json.dumps(err, ensure_ascii=False))
        return err

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Seedream Image Generation API Wrapper")
    parser.add_argument("--prompt", type=str, required=True, help="Image generation prompt")
    parser.add_argument("--model", type=str, help="Model endpoint id")
    parser.add_argument("--size", type=str, help="Image size")
    parser.add_argument("--watermark", type=str, help="Add watermark (true/false)")
    parser.add_argument("--optimize_prompt_options", type=str, help="JSON string for prompt optimization options")
    parser.add_argument("--tools", type=str, help="JSON string for tools list, e.g. '[{\"type\": \"web_search\"}]'")
    parser.add_argument("--output_format", type=str, help="Output image format (png, jpeg)")
    parser.add_argument("--sequential_image_generation", type=str, help="Sequential image generation mode (e.g., 'auto')")
    parser.add_argument("--download_dir", type=str, help="Directory to save images")
    parser.add_argument("--image", type=str, help="Local image path or JSON string list of local image paths, e.g. '/a.png' or '[\"/a.png\",\"/b.jpg\"]'")
    parser.add_argument("--response_format", type=str, help="Response format for edits (url or b64_json)")
    args = parser.parse_args()
    
    watermark_val = None
    if args.watermark is not None:
        watermark_val = args.watermark.lower() == "true"
        
    optimize_prompt_val = None
    if args.optimize_prompt_options is not None:
        try:
            optimize_prompt_val = json.loads(args.optimize_prompt_options)
        except json.JSONDecodeError:
            print("Error: --optimize_prompt_options must be a valid JSON string.")
            sys.exit(1)

    tools_val = None
    if args.tools is not None:
        try:
            tools_val = json.loads(args.tools)
        except json.JSONDecodeError:
            print("Error: --tools must be a valid JSON string.")
            sys.exit(1)

    image_val = None
    if args.image is not None:
        if args.image.strip().startswith("["):
            try:
                image_val = json.loads(args.image)
                if not isinstance(image_val, list):
                    print("Error: --image must be a local path or a JSON list of paths.")
                    sys.exit(1)
            except json.JSONDecodeError:
                print("Error: --image must be a local path or a valid JSON string list.")
                sys.exit(1)
        else:
            image_val = args.image

    res = generate_image(
        prompt=args.prompt,
        model=args.model,
        size=args.size,
        watermark=watermark_val,
        optimize_prompt_options=optimize_prompt_val,
        tools=tools_val,
        output_format=args.output_format,
        sequential_image_generation=args.sequential_image_generation,
        download_dir=args.download_dir,
        image=image_val,
        response_format=args.response_format
    )
    print(json.dumps(res, indent=2))
